File: strings/12_valid_brackets/valid_brackets.py
from typing import List


class Solution:
    def is_valid(self, s: str) -> bool:
        if not self._check_constraints(s):
            raise ValueError("Constraints violated")
        if len(s) % 2 != 0:
            return False
        stck: List[str] = []
        for ch in s:
            if ch == '(' or ch == '{' or ch == '[':
                stck.append(ch)
            else:
                if len(stck) == 0:
                    return False
                if ch == ')' and not stck.pop() == '(':
                    return False
                elif ch == ']' and not stck.pop() == '[':
                    return False
                elif ch == '}' and not stck.pop() == '{':
                    return False

        return True if len(stck) == 0 else False

    def _check_constraints(self, s: str) -> bool:
        if len(s) < 1 or len(s) > 10**4:
            return False
        for ch in s:
            if not (ch == '(' or ch == ')' or ch == '[' or ch == ']' or ch == '{' or ch == '}'):
                return False
        return True

File: strings/12_valid_brackets/valid_brackets.cpp
#include <iostream>
#include <cmath>
#include <stdexcept>
#include <cstdint>
#include <string>
#include <vector>
#include <stack>

using namespace std;


class Solution {
public:
    bool isValid(string s) {
        if (!checkConstraints(s)) {
            throw std::runtime_error("Constraints violated");
        }
        if (s.size() % 2 != 0) return false;
        stack<char, vector<char>> stck;
        for (auto& ch : s) {
            if (ch == '(' || ch == '{' || ch == '[') {
                stck.push(ch);
            } else {
                if (stck.empty()) return false;
                if (ch == ')') {
                    if (stck.top() == '(') stck.pop();
                    else return false;
                } else if (ch == '}') {
                    if (stck.top() == '{') stck.pop();
                    else return false;                 
                } else {
                    if (stck.top() == '[') stck.pop();
                    else return false;  
                }
            }
        }
        return stck.empty() ? true : false;
    }

private:
    bool checkConstraints(const string& s) {
        if (s.size() < 1 || s.size() > pow(10, 4)) return false;
        for (const char& ch : s) {
            if (!(ch == '(' || ch == ')' || ch == '{' || ch == '}' || ch == '[' || ch == ']')) return false;
        }
        return true;
    }
};